Enhance evaluate.py to output JSON file for benchmark

hadusam opened the pull request:
<!--- 👉👉👉 Provide a general summary of your changes. 👈👈👈 -->

## Motivation and Context
<!--- 💭💡💭 Why is this change required? What problem does it solve? 💭💡💭 -->
<!--- 🔗 If it fixes an open issue, please link to the issue here. 🔗 -->
Current evaluate.py creates a binary file of evaluation result for tensorboard.
We need to get a readable file of evaluation results for humans, so I enhanced it to create JSON file.
This enhancement is also used for the benchmark of our models.

## Description
<!--- 📝🎯📝 Describe your changes in detail. 📝🎯📝 -->

## How has this been tested?
<!--- 🙆👌🙆 Please describe in detail how you tested your changes. 🙆👌🙆 -->
<!--- Include details of your testing environment, details of your manual tests, snippet code or commands of your manual tests, table of experiments results metrics, tests ran to see how your change affects other areas of the code, etc. -->
<!--- If you have experiments report documents please link to here. -->
```
$ docker-compose run --rm blueoil python lmnet/executor/evaluate.py -i train_20191216084759

... some warnings

----------------- config value --------------------
{'BATCH_SIZE': 1,
 'CLASSES': ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9'],
 'DATASET': {'AUGMENTOR': [Blur({'min_value': 0, 'max_value': 1})],
             'BATCH_SIZE': 1,
             'DATA_FORMAT': 'NHWC',
             'ENABLE_PREFETCH': True,
             'PRE_PROCESSOR': [Resize({'size': [128, 128]}), DivideBy255({})]},
 'DATASET_CLASS': <class 'abc.DATASET_CLASS'>,
 'DATA_FORMAT': 'NHWC',
 'IMAGE_SIZE': [128, 128],
 'IS_DEBUG': False,
 'IS_PRETRAIN': False,
 'KEEP_CHECKPOINT_MAX': 5,
 'MAX_EPOCHS': 1,
 'NETWORK': {'ACTIVATION_QUANTIZER': <function linear_mid_tread_half_quantizer at 0x7f168695b9d8>,
             'ACTIVATION_QUANTIZER_KWARGS': {'bit': 2, 'max_value': 2},
             'BATCH_SIZE': 1,
             'DATA_FORMAT': 'NHWC',
             'IMAGE_SIZE': [128, 128],
             'LEARNING_RATE_FUNC': None,
             'LEARNING_RATE_KWARGS': None,
             'OPTIMIZER_CLASS': <class 'tensorflow.python.training.adam.AdamOptimizer'>,
             'OPTIMIZER_KWARGS': {'learning_rate': 0.001},
             'WEIGHT_DECAY_RATE': 0.0005,
             'WEIGHT_QUANTIZER': <function binary_mean_scaling_quantizer at 0x7f168695b8c8>,
             'WEIGHT_QUANTIZER_KWARGS': {}},
 'NETWORK_CLASS': <class 'lmnet.networks.classification.lmnet_v1.LmnetV1Quantize'>,
 'POST_PROCESSOR': None,
 'PRETRAIN_DIR': '',
 'PRETRAIN_FILE': '',
 'PRETRAIN_VARS': [],
 'PRE_PROCESSOR': [Resize({'size': [128, 128]}), DivideBy255({})],
 'SAVE_CHECKPOINT_STEPS': 1000,
 'SUMMARISE_STEPS': 100,
 'TASK': <Tasks.CLASSIFICATION: 'IMAGE.CLASSIFICATION'>,
 'TEST_STEPS': 1000}
----------------- config value --------------------
restore_path: /home/blueoil/saved/train_20191216084759/checkpoints/save.ckpt-27
ENABLE prefetch

... some warnings

INFO:tensorflow:Restoring parameters from /home/blueoil/saved/train_20191216084759/checkpoints/save.ckpt-27
INFO:tensorflow:Restoring parameters from /home/blueoil/saved/train_20191216084759/checkpoints/save.ckpt-27
INFO:__main__:test_step_size3
INFO:__main__:test_step0
INFO:__main__:test_step1
INFO:__main__:test_step2
INFO:lmnet.utils.predict_output.writer:save json: /home/blueoil/saved/train_20191216084759/evaluate/json/27.json
```

```
$ cat saved/train_20191216084759/evaluate/json/27.json
{
    "last_step": 27,
    "metrics": [
        {
            "name": "accuracy",
            "value": 0.0
        },
        {
            "name": "accuracy_top3",
            "value": 0.0
        },
        {
            "name": "accuracy_top5",
            "value": 0.3333333432674408
        }
    ]
}
```
## Screenshots (if appropriate):

## Types of changes
<!--- 🏁🎌🏁 What types of changes does your code introduce? Put an `x` in all the boxes that apply: 🏁🎌🏁 -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:
<!--- ✅✅✅ Go over all the following points, and put an `x` in all the boxes that apply. ✅✅✅ -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

<!--- TODO(wakisaka): After decide our code style, add this.
- [ ] My code follows the code style of this project.
-->
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

iizukak commented on lmnet/executor/evaluate.py:
```diff
@@ -25,6 +27,9 @@
 from lmnet.datasets.tfds import TFDSClassification, TFDSObjectDetection
 from lmnet.utils import config as config_util
 from lmnet.utils import executor, module_loader
+from lmnet.utils.predict_output.writer import save_json
```
[just comment] The metrics is `predict_output` 🤔

hadusam commented on lmnet/executor/evaluate.py:
```diff
@@ -44,15 +49,18 @@ def setup_dataset(config, subset, seed):
 
 
-def evaluate(config, restore_path):
+def evaluate(config, restore_path, output_dir):
```
Yes! Please see [search result](https://github.com/blue-oil/blueoil/search?utf8=%E2%9C%93&q=evaluate&type=).

hadusam commented on lmnet/executor/evaluate.py:
```diff
@@ -140,6 +148,12 @@ def evaluate(config, restore_path):
 
+    metrics_dict = {
+        'last_step': int(last_step),
+        'metrics': dict([[k, float(v)] for k, v in zip(list(metrics_ops_dict.keys()), metrics_values)]),
```
@iizukak 
Oh, thanks! I fixed it 👍

kchygoe commented on lmnet/executor/evaluate.py:
```diff
@@ -25,6 +27,9 @@
 from lmnet.utils import executor, module_loader
+from lmnet.utils.predict_output.writer import save_json
+
+logging.basicConfig(level=logging.INFO)
```
`logger = logging.getLogger(__name__)`

kchygoe replied:
and use `logger`.

hadusam replied:
Thanks! @kchygoe 
I changed it to use logger and I also fixed `lmnet.utils.predict_output.writer` as same.
The log is changed to show module name (instead of `root`) as following.
```
INFO:__main__:test_step_size3
INFO:__main__:test_step0
INFO:__main__:test_step1
INFO:__main__:test_step2
INFO:lmnet.utils.predict_output.writer:save json: /home/blueoil/saved/train_20191216084759/evaluate/json/27.json
```

hadusam commented:
@tsawada 
Please review the readability of this PR, thanks! 👍 

@suttang 
Please check and confirm the JSON format 🙇

tsawada commented on lmnet/executor/evaluate.py:
```diff
@@ -140,6 +149,12 @@ def evaluate(config, restore_path):
 
+    metrics_dict = {
+        'last_step': int(last_step),
+        'metrics': [{'name': k, 'value': float(v)} for k, v in zip(list(metrics_ops_dict.keys()), metrics_values)],
```
Order of `list(dict.keys())` is not predictable. if `metrics_ops_dict` is a dict, I'm not sure if that will match with `metrics_values`.

hadusam replied:
@tsawada 
Thanks!
I agree with you for readability but this issue is not only this PR, but we also have the same implementation on `lmnet/executor/train.py` and need to change another logics (lmnet/lmnet/utils.py).
I created the issue #704, better to solve another PR. Could you please allow let this is?


> that will match with metrics_values

In this implementation, only using `metrics_ops_dict.values()` and `metrics_ops_dict.keys()` and these orders is the same if there are no changes in `dict` objects. But it is not so good because this is implicit.

https://stackoverflow.com/questions/835092/python-dictionary-are-keys-and-values-always-the-same-order

hadusam replied:
@tsawada 
Thanks!
I agree with you for readability but this issue is not only this PR, but we also have the same implementation on `lmnet/executor/train.py` and need to change another logics (lmnet/lmnet/utils.py).
I created the issue #704, better to solve another PR. Could you please allow let this is?


> that will match with metrics_values

In this implementation, only using `metrics_ops_dict.values()` and `metrics_ops_dict.keys()` and these orders is the same if there are no changes in `dict` objects. But it is not so good because this is implicit.

https://stackoverflow.com/questions/835092/python-dictionary-are-keys-and-values-always-the-same-order

tsawada replied:
Can you add a TODO comment that this needs to be fixed?

hadusam commented:
@tsawada 
> Do we need to stash logging change in this diff?

I would like to get the information about the saved path of the JSON file as follows.
```
INFO:lmnet.utils.predict_output.writer:save json: /home/blueoil/saved/train_20191216084759/evaluate/json/27.json
```

`lmnet.utils.predict_output.writer` is using `logging` and need to change log level to INFO to get saved path information.
So I added `import logging` and `logging.basicConfig(level=logging.INFO)`, I need it.

But current `evaluate.py` is using print debugging, following #417, I added small changes to use `logger` instead of `print` in this timing (when adding `import logging`).
See also: https://github.com/blue-oil/blueoil/pull/682#discussion_r359190144
I think these are small changes, and not noisy for review, but if it is better to separate them, I will.

hadusam commented:
@tsawada 
I added some comments, please re-review this 🙇

hadusam commented:
@tsawada
I separated the changing `logging` part, please review again.

hadusam commented:
@tsawada Added TODO comment 👍

hadusam commented:
/ready
